packages/flutter/lib/src/widgets/focus_scope.dart

```diff
@@ -106,38 +119,70 @@ class Focus extends StatefulWidget {
     this.focusNode,
     this.autofocus = false,
     this.onFocusChange,
-    this.onKey,
-    this.onKeyEvent,
-    this.debugLabel,
-    this.canRequestFocus,
-    this.descendantsAreFocusable = true,
-    this.skipTraversal,
+    FocusOnKeyCallback? onKey,
+    FocusOnKeyEventCallback? onKeyEvent,
+    String? debugLabel,
+    bool? canRequestFocus,
+    bool descendantsAreFocusable = true,
+    bool? skipTraversal,
     this.includeSemantics = true,
-  })  : assert(child != null),
+  })  : _usingExternalFocus = false,
+        _onKeyEvent = onKeyEvent,
+        _onKey = onKey,
+        _skipTraversal = skipTraversal,
+        _canRequestFocus = canRequestFocus,
+        _descendantsAreFocusable = descendantsAreFocusable,
+        _debugLabel = debugLabel,
+        assert(child != null),
         assert(autofocus != null),
         assert(descendantsAreFocusable != null),
         assert(includeSemantics != null),
         super(key: key);
 
+  /// Creates a Focus widget that uses the given [focusNode] as the source of
+  /// truth for attributes on the node, rather than the attributes of this widget.
+  const Focus.withExternalFocusNode({
+    Key? key,
+    required this.child,
+    required FocusNode this.focusNode,
+    this.autofocus = false,
+    this.onFocusChange,
+    this.includeSemantics = true,
+  })  : _usingExternalFocus = true,
+        _onKeyEvent = null,
+        _onKey = null,
+        _skipTraversal = null,
+        _canRequestFocus = null,
+        _descendantsAreFocusable = true,
+        _debugLabel = null,
+        assert(child != null),
+        assert(autofocus != null),
+        assert(includeSemantics != null),
+        super(key: key);
+
+  // Indicates whether the widget's focusNode attributes should have priority
+  // when then widget is updated.
+  final bool _usingExternalFocus;
+
   /// A debug label for this widget.
   ///
   /// Not used for anything except to be printed in the diagnostic output from
-  /// [toString] or [toStringDeep]. Also unused if a [focusNode] is provided,
-  /// since that node can have its own [FocusNode.debugLabel].
+  /// [toString] or [toStringDeep].
   ///
   /// To get a string with the entire tree, call [debugDescribeFocusTree]. To
   /// print it to the console call [debugDumpFocusTree].
   ///
   /// Defaults to null.
-  final String? debugLabel;
+  String? get debugLabel => _usingExternalFocus ? focusNode!.debugLabel : _debugLabel ?? focusNode?.debugLabel;
```

Comment:
super nit: maybe it will be cleaner if the external focus thing was made a private subclass of `Focus` (and no `!` will be needed I think), and redirect its constructor using a const factory constructor?

Reply:
OK, I'll try that and see how it looks.

Reply:
OK, I did this for both the `Focus` and `FocusScope` classes. Looks pretty good.  I still had to keep an accessor for whether or not it was using an external node, since the state class uses that to see if it should update the focus node attributes or not.